### PR TITLE
Add time constraint for progressive rendering

### LINKED ISSUE #1508: Add progressive rendering time limit
Add an option in studio to use time as a rendering constraint instead of samples per pixel.

### PR DESCRIPTION
**Pull request by mororo250**
Fix #1508

### CODE REVIEW

**Comment by mororo250** [comment]
Screenshot (7)

File: src/appleseed/renderer/kernel/rendering/progressive/progressiveframerenderer.cpp [hunk 1]
```diff
@@ -743,6 +761,15 @@ Dictionary ProgressiveFrameRendererFactory::get_params_metadata()
...
             .insert("help", "Maximum number of average samples per pixel"));
 
+    metadata.dictionaries().insert(
+        "time_limit",
+        Dictionary()
+            .insert("type", "int")
+            .insert("default", "60")
+            .insert("unlimited", "true")
+            .insert("label", "Time Limit:")
+            .insert("help", "Maximum time rendering"));
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/progressive/progressiveframerenderer.cpp]: "Maximum rendering time"

File: src/appleseed/renderer/kernel/rendering/progressive/progressiveframerenderer.cpp [hunk 2]
```diff
@@ -366,6 +380,7 @@ namespace
             const SamplingContext::Mode m_sampling_mode;
             const size_t                m_thread_count;       // number of rendering threads
             const uint64                m_max_average_spp;    // maximum average number of samples to compute per pixel
+            const uint64                m_time_limit;         // maximum time rendering
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/progressive/progressiveframerenderer.cpp]: Better: `maximum rendering time in seconds`

File: src/appleseed/renderer/kernel/rendering/progressive/progressiveframerenderer.cpp [hunk 3]
```diff
@@ -222,7 +231,8 @@ namespace
 
         bool is_rendering() const override
         {
-            return m_job_queue.has_scheduled_or_running_jobs();
+            m_stopwatch.measure();
+            return (m_job_queue.has_scheduled_or_running_jobs() && m_params.m_time_limit > m_stopwatch.get_seconds());
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/progressive/progressiveframerenderer.cpp]: We need to revert this: even if the maximum rendering time has been reached, if there are still running jobs, this method should return true. If it lies and returns false, it's going to cause bugs of all sorts.

The proper way to implement this is to use a `IRendererController`. The host application already passes one, maybe we need to wrap it into another one that implements time limiting.

File: src/appleseed/renderer/kernel/rendering/progressive/progressiveframerenderer.cpp [hunk 4]
```diff
@@ -206,6 +210,11 @@ namespace
                 m_params.m_max_average_spp == numeric_limits<uint64>::max()
                     ? "unlimited"
                     : pretty_uint(m_params.m_max_average_spp).c_str(),
+                m_params.m_time_limit == numeric_limits<uint64>::max()
+                ? "unlimited"
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/progressive/progressiveframerenderer.cpp]: Fix indentation.

File: src/appleseed/renderer/kernel/rendering/progressive/progressiveframerenderer.cpp [hunk 5]
```diff
@@ -206,6 +210,11 @@ namespace
...
+                : std::string(pretty_uint(hours) + "h "
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/progressive/progressiveframerenderer.cpp]: Use `pretty_time()` instead.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 6]
```diff
@@ -432,19 +435,37 @@ namespace
...
+    };
+
+    class FinalGeneralSettingsPanel
+      : public GeneralSettingsPanel
+    {
+        Q_OBJECT
+
+      public:
+        explicit FinalGeneralSettingsPanel(const Configuration& config, QWidget* parent = nullptr)
+            : GeneralSettingsPanel(parent)
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Just two spaces (half indent) here.

This idiosyncracy will go away when we switch to clang-format (after the next release)...

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 7]
```diff
@@ -455,6 +476,80 @@ namespace
...
     };
 
+    class InteractiveGeneralSettingsPanel
+      : public GeneralSettingsPanel
+    {
+        Q_OBJECT
+
+      public:
+        explicit InteractiveGeneralSettingsPanel(const Configuration& config, QWidget* parent = nullptr)
+            : GeneralSettingsPanel(parent)
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Half indent here too.

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 8]
```diff
@@ -455,6 +476,80 @@ namespace
...
+            QCheckBox* unlimited_time = create_checkbox("unlimited_time", "Unlimited");
+            parent->addRow("Time Limit:", create_horizontal_group(hours, minutes, seconds, unlimited_time));
+            connect(unlimited_time, SIGNAL(toggled(bool)), hours, SLOT(setDisabled(bool)));
+            connect(unlimited_time, SIGNAL(toggled(bool)), minutes, SLOT(setDisabled(bool)));
+            connect(unlimited_time, SIGNAL(toggled(bool)), seconds, SLOT(setDisabled(bool)));
+        }
+
+        void load_time_limit(const Configuration& config)
+        {
+            constexpr int DefaultSeconds = 0;
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: 👍

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 9]
```diff
@@ -455,6 +476,80 @@ namespace
...
+        }
+
+        void save_time_limit(Configuration & config) const
+        {
+            if (get_widget<bool>("unlimited_time"))
+                config.get_parameters().remove_path("progressive_frame_renderer.time_limit");
+            else
+            {
+                // Transform to seconds.
+                unsigned int hours = get_widget<unsigned int>("hours");
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Make all variables const.

**Comment by mororo250** [comment]
Thanks for the review.
I addressed your comments.

File: src/appleseed/renderer/kernel/rendering/renderercontrollercollection.cpp [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+{
+    for (auto i : impl->m_renderer_controller)
+        i->on_progress();
+}
+
+IRendererController::Status RendererControllerCollection::get_status() const
+{
+    IRendererController::Status status = IRendererController::Status::ContinueRendering;
+    for (auto i : impl->m_renderer_controller)
+        status = i->get_status() > status ? i->get_status() : status;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/renderercontrollercollection.cpp]: That's an interesting choice. You assume that there's an ordering in the statuses but there isn't really one. I think we could live with simply returning the first status that isn't `ContinueRendering`. What do you think?

File: src/appleseed.studio/mainwindow/rendering/renderingmanager.cpp [hunk 11]
```diff
@@ -199,7 +199,16 @@ void RenderingManager::start_rendering(
...
     m_render_tab = render_tab;
 
-    m_render_tab->get_render_widget()->start_render();
+    m_renderer_controller_collection.reset(new RendererControllerCollection());
+
+    m_renderer_controller_collection->insert(&m_renderer_controller);
+
+    if(params.get_required<string>("frame_renderer", "generic") == "progressive")
+    {
+        TimedRendererController* time_controller = new TimedRendererController(
```
[the remark this answers is not in the document]
Comment (mororo250) [reply src/appleseed.studio/mainwindow/rendering/renderingmanager.cpp]: This was fixed in the last commit.

File: src/appleseed/renderer/kernel/rendering/renderercontrollercollection.cpp [hunk 12]
```diff
@@ -0,0 +1,118 @@
+
+//
+// This source file is part of appleseed.
+// Visit https://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright(c) 2019 João Marcos Costa, The appleseedhq Organization
```
[the remark this answers is not in the document]
Comment (mororo250) [reply src/appleseed/renderer/kernel/rendering/renderercontrollercollection.cpp]: Hahaha
okay I will change the ones of my other pr as well.

File: src/appleseed.studio/mainwindow/rendering/renderingmanager.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (dictoon) [reply src/appleseed.studio/mainwindow/rendering/renderingmanager.cpp]: Ah right, I missed it, sorry.

File: src/appleseed/renderer/kernel/rendering/renderercontrollercollection.cpp (hunk 10 above)
Comment (mororo250) [reply src/appleseed/renderer/kernel/rendering/renderercontrollercollection.cpp]: Agreed!

File: src/appleseed.studio/mainwindow/rendering/renderingmanager.cpp [hunk 13]
```diff
@@ -199,7 +199,16 @@ void RenderingManager::start_rendering(
     m_rendering_mode = rendering_mode;
     m_render_tab = render_tab;
 
-    m_render_tab->get_render_widget()->start_render();
+    m_renderer_controller_collection.reset(new RendererControllerCollection());
+
+    m_renderer_controller_collection->insert(&m_renderer_controller);
+
+    if(params.get_required<string>("frame_renderer", "generic") == "progressive")
```
[the remark this answers is not in the document]
Comment (mororo250) [reply src/appleseed.studio/mainwindow/rendering/renderingmanager.cpp]: I like the idea, but I am not sure if I got how it would work.

**Comment by mororo250** [comment]
I think is done.
We should create an issue for
> 3. We can go with the solution you implemented, but looking at the code it now occurs to me that a more general solution would be to have frame renderers always implement the `IRendererController` interface so that they can decide by themselves when to stop or pause rendering, or have them return a renderer controller via some method. This could be done in a future PR to improve and generalize the design. What do you think?

**Comment by dictoon** [comment]
Perfect, thank you!

I agree about the new issue, I'll create it now.

**Pull request merged**